### ISSUE TITLE
Provider not inherited by nested module call

### ISSUE THREAD

**Issue by pikeas** [hashicorp/terraform#3931]
I have a main.tf which calls into Module1.

Module1's main calls into Module2 (which is in a sub-directory, but I don't think that matters).

If Module1 defines at least one resource, everything is fine. If Module1 defines no resources, because it consists only of variables, outputs, and calls to other modules, those modules will fail to inherit provider settings.

Specifically, my `terraform plan` fails with: module.kubernetes.module.master.provider.aws: "region": required field is not set. To clarify the names, this is a Kubernetes module with a master sub-module. Adding any resource into `kubernetes/main.tf` eliminates the error.

**Comment by jen20** [comment]
Hi @pikeas! Thanks for reporting this. I'm trying to reproduce this according to your description, but currently can't. Do you have a set of configuration which demonstrates the issue?

The configuration I'm testing with is:

main.tf:
```hcl
provider "aws" {
    region = "us-east-1"
}

module "inner" {
    source = "./inner"
}

output "vpc_id" {
    value = "${module.inner.vpc_id}"
}
```

inner/main.tf:
```hcl
provider "aws" {
    region = "us-west-2"
}

resource "aws_vpc" "main" {
    cidr_block = "10.0.0.0/16"
}

output "vpc_id" {
    value = "${aws_vpc.main.id}"
}
```

This behaves as I'd expect, creating the VPC in the `us-east-1` region. Removing the region from the inner module yields the same result.

**Comment by jen20** [comment]
Seems related to #2832

**Comment by jen20** [comment]
@pikeas - I'm going to close this issue in favour of tracking it on #2832 since we have a repro there. I'll try to remember to update this one when it gets fixed, but it might be worth you watching #2832 as well.
